File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.js
function solution(files) {
    files.sort((a, b) => {
        const [aHead, aNumber] = parseFileName(a);
        const [bHead, bNumber] = parseFileName(b);

        return aHead.localeCompare(bHead) || aNumber - bNumber;
    });
    return files;
}

function parseFileName(file) {
    const match = file.match(/^([^\d]+)(\d{1,5})/);
    const head = match ? match[1].toLowerCase() : "";
    const number = match ? parseInt(match[2], 10) : 0;
    return [head, number];
}

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.java
import java.util.*;
import java.util.regex.*;

public class Solution {
    public String[] solution(String[] files) {
        Pattern p = Pattern.compile("^([\\D]+)(\\d{1,5})");
        
        Arrays.sort(files, new Comparator<String>() {
            @Override
            public int compare(String a, String b) {
                Matcher mA = p.matcher(a.toLowerCase());
                Matcher mB = p.matcher(b.toLowerCase());

                mA.find();
                mB.find();

                if (!mA.group(1).equals(mB.group(1))) {
                    return mA.group(1).compareTo(mB.group(1));
                } else {
                    return Integer.parseInt(mA.group(2)) - Integer.parseInt(mB.group(2));
                }
            }
        });
        return files;
    }
}